File: matlab/sensitivity_analysis.m
% sensitivity_analysis.m
% Three-phase sensitivity analysis of the cyst-counting parameters:
%   Phase 1 – Morris screening (Campolongo et al. 2007)
%   Phase 2 – One-at-a-time (OAT) analysis, all 6 parameters ± 30 %
%   Phase 3 – 2-D parameter interaction for the top 2 parameters
%
% Requires: run_optimization.m to have been run (reads Optimization set masks).

clear
clc

scriptDir = fileparts(mfilename('fullpath'));
addpath(scriptDir);

%% Configuration
optimalParams = [4.27269938837114, 18.2632731580751, 42144.3492985436, ...
                 14.9647743085524, 0.111299008136904, 0.918315428393029];
paramNames    = {'minPeakSepSmall', 'minPeakSepLarge', 'largeCystVolThresh', ...
                 'maxPeakDistMerge', 'thresholdFraction', 'gaussianSigma'};

dataDir  = fullfile(scriptDir, '..', 'data', 'Optimization set');
maskDir  = fullfile(dataDir, 'Cyst masks');
xlsFile  = fullfile(dataDir, 'O1_count.xlsx');
outputDir = fullfile(scriptDir, '..', 'results', 'sensitivity_analysis');
if ~exist(outputDir, 'dir'), mkdir(outputDir); end

if ~isfile(xlsFile)
    error('Ground-truth file not found:\n  %s\nPlace the Zenodo data inside data/.', xlsFile);
end

T   = readtable(xlsFile);
n   = height(T);
gt_masks = cell(n, 1);
gt_am    = zeros(1, n);
for i = 1:n
    id          = T{i,1}{1};
    gt_am(i)    = T{i,2};
    gt_masks{i} = fullfile(maskDir, sampleIDtoFilename(id));
end

lowerBounds = optimalParams * 0.7;
upperBounds = optimalParams * 1.3;
BATCH_SIZE  = 8;

% Parallel pool
p = gcp('nocreate');
if isempty(p)
    parpool('Processes', 8);
end

%% Phase 1 – Morris screening
fprintf('\n=== Phase 1: Morris screening ===\n');
tic_morris = tic;

r     = 20;
M     = 1000;
p_dim = length(optimalParams);

fprintf('Generating %d candidate trajectories...\n', M);
candidates = cell(M, 1);
num_levels = 6;
delta      = num_levels / (2 * (num_levels - 1));
for m = 1:M
    candidates{m} = morrisTrajectory(p_dim, delta);
end

fprintf('Computing pairwise distances...\n');
distMat = zeros(M, M);
for m = 1:M
    for l = m+1:M
        distMat(m,l) = trajectoryDistance(candidates{m}, candidates{l});
        distMat(l,m) = distMat(m,l);
    end
    if mod(m, 200) == 0, fprintf('  %d / %d\n', m, M); end
end

selectedIdx = selectMaxSpread(distMat, r);

% Build parameter sets
allParams = cell(r * (p_dim + 1), 1);
for idx = 1:r
    B = candidates{selectedIdx(idx)};
    for i = 1:(p_dim + 1)
        allParams{(idx-1)*(p_dim+1) + i} = lowerBounds + B(i,:) .* (upperBounds - lowerBounds);
    end
end

fprintf('Evaluating %d parameter sets...\n', length(allParams));
allScores = evaluateInBatches(allParams, gt_masks, gt_am, BATCH_SIZE);

% Compute elementary effects
EE = zeros(r, p_dim);
for traj = 1:r
    base  = (traj-1)*(p_dim+1);
    sTraj = allScores((base+1):(base+p_dim+1));
    B     = candidates{selectedIdx(traj)};
    for step = 1:p_dim
        dNorm = B(step+1,:) - B(step,:);
        changed = find(abs(dNorm) > 1e-6);
        if ~isempty(changed)
            EE(traj, changed) = (sTraj(step+1) - sTraj(step)) / dNorm(changed);
        end
    end
end

morrisRes.mu_star = mean(abs(EE), 1)';
morrisRes.mu      = mean(EE, 1)';
morrisRes.sigma   = std(EE, 0, 1)';

fprintf('Morris screening completed in %.1f min\n', toc(tic_morris)/60);

% Morris plot
fig1 = figure('Position', [100,100,800,600], 'Visible', 'off');
scatter(morrisRes.mu_star, morrisRes.sigma, 120, 'filled', 'MarkerFaceAlpha', 0.7);
xlabel('\mu*', 'FontSize', 12, 'FontWeight', 'bold');
ylabel('\sigma', 'FontSize', 12, 'FontWeight', 'bold');
title('Morris screening', 'FontSize', 13, 'FontWeight', 'bold');
grid on; hold on;
for i = 1:length(paramNames)
    text(morrisRes.mu_star(i)*1.04, morrisRes.sigma(i)*1.04, paramNames{i}, 'FontSize', 10);
end
xline(median(morrisRes.mu_star), '--r', 'LineWidth', 1.5);
yline(median(morrisRes.sigma),   '--b', 'LineWidth', 1.5);
saveas(fig1, fullfile(outputDir, 'morris_screening.png'));
close(fig1);

importance_threshold = median(morrisRes.mu_star);
[~, sortIdx] = sort(morrisRes.mu_star, 'descend');

fprintf('\n%-25s %10s %10s %10s\n', 'Parameter', 'mu*', 'mu', 'sigma');
fprintf('%s\n', repmat('-', 1, 60));
for i = 1:length(paramNames)
    k = sortIdx(i);
    star = '';
    if morrisRes.mu_star(k) > importance_threshold, star = ' *'; end
    fprintf('%-25s %10.4f %10.4f %10.4f%s\n', paramNames{k}, ...
            morrisRes.mu_star(k), morrisRes.mu(k), morrisRes.sigma(k), star);
end

%% Phase 2 – OAT analysis
fprintf('\n=== Phase 2: One-at-a-time analysis ===\n');
tic_oat = tic;

variations  = linspace(-0.3, 0.3, 21);
nVar        = length(variations);
oat_order   = sortIdx;

allOATParams = cell(length(oat_order) * nVar, 1);
paramMap     = zeros(length(oat_order) * nVar, 2);
idx = 1;
for i = 1:length(oat_order)
    for v = 1:nVar
        tp = optimalParams;
        tp(oat_order(i)) = optimalParams(oat_order(i)) * (1 + variations(v));
        allOATParams{idx} = tp;
        paramMap(idx, :)  = [i, v];
        idx = idx + 1;
    end
end

[allOATScores, allOATMAEs] = evaluateInBatchesFull(allOATParams, gt_masks, gt_am, BATCH_SIZE);
elapsed_oat = toc(tic_oat);

oatResults = cell(length(oat_order), 1);
for i = 1:length(oat_order)
    mask = paramMap(:,1) == i;
    oatResults{i} = struct('param', paramNames{oat_order(i)}, ...
                           'variations', variations, ...
                           'scores', allOATScores(mask), ...
                           'MAEs', allOATMAEs(mask));
end

% OAT plots
fig2 = figure('Position', [100,100,1600,900], 'Visible', 'off');
for i = 1:length(oat_order)
    subplot(2,3,i);
    d = oatResults{i};
    yyaxis left
    plot(d.variations*100, d.scores, 'o-', 'LineWidth', 2, 'MarkerSize', 7, 'Color', [0.2 0.4 0.8]);
    ylabel('MAPE (%)', 'FontSize', 10, 'FontWeight', 'bold');
    yyaxis right
    plot(d.variations*100, d.MAEs, 's--', 'LineWidth', 2, 'MarkerSize', 7, 'Color', [0.8 0.3 0.2]);
    ylabel('MAE (cysts)', 'FontSize', 10, 'FontWeight', 'bold');
    xlabel('Variation (%)', 'FontSize', 10);
    title(sprintf('%s (rank %d)', d.param, i), 'FontSize', 11, 'FontWeight', 'bold');
    xline(0, 'r--', 'LineWidth', 2);
    grid on;
end
sgtitle('One-at-a-time sensitivity analysis', 'FontSize', 14, 'FontWeight', 'bold');
saveas(fig2, fullfile(outputDir, 'oat_analysis.png'));
close(fig2);
fprintf('OAT completed in %.1f min\n', elapsed_oat/60);

%% Phase 3 – 2-D interaction
fprintf('\n=== Phase 3: 2-D parameter interaction ===\n');
tic_2d = tic;

p1_idx    = oat_order(1);
p2_idx    = oat_order(2);
grid_size = 9;
grid_vals = linspace(-0.25, 0.25, grid_size);

all2DParams = cell(grid_size^2, 1);
gridMap     = zeros(grid_size^2, 2);
idx = 1;
for i = 1:grid_size
    for j = 1:grid_size
        tp = optimalParams;
        tp(p1_idx) = optimalParams(p1_idx) * (1 + grid_vals(i));
        tp(p2_idx) = optimalParams(p2_idx) * (1 + grid_vals(j));
        all2DParams{idx} = tp;
        gridMap(idx, :)  = [i, j];
        idx = idx + 1;
    end
end

[scores2D, maes2D] = evaluateInBatchesFull(all2DParams, gt_masks, gt_am, BATCH_SIZE);

intScores = zeros(grid_size); intMAE = zeros(grid_size);
for k = 1:length(all2DParams)
    intScores(gridMap(k,1), gridMap(k,2)) = scores2D(k);
    intMAE(gridMap(k,1), gridMap(k,2))    = maes2D(k);
end

elapsed_2d = toc(tic_2d);
fprintf('2-D interaction completed in %.1f min\n', elapsed_2d/60);

fig3 = figure('Position', [100,100,1300,550], 'Visible', 'off');
subplot(1,2,1)
imagesc(grid_vals*100, grid_vals*100, intScores); colorbar; colormap(jet);
xlabel(sprintf('%s variation (%%)', paramNames{p2_idx}), 'FontWeight', 'bold');
ylabel(sprintf('%s variation (%%)', paramNames{p1_idx}), 'FontWeight', 'bold');
title('MAPE landscape (%)'); axis square;
hold on; plot(0, 0, 'wx', 'MarkerSize', 18, 'LineWidth', 3);

subplot(1,2,2)
imagesc(grid_vals*100, grid_vals*100, intMAE); colorbar; colormap(jet);
xlabel(sprintf('%s variation (%%)', paramNames{p2_idx}), 'FontWeight', 'bold');
ylabel(sprintf('%s variation (%%)', paramNames{p1_idx}), 'FontWeight', 'bold');
title('MAE landscape (cysts)'); axis square;
hold on; plot(0, 0, 'wx', 'MarkerSize', 18, 'LineWidth', 3);

sgtitle(sprintf('Parameter interaction: %s vs %s', paramNames{p1_idx}, paramNames{p2_idx}), ...
        'FontSize', 13, 'FontWeight', 'bold');
saveas(fig3, fullfile(outputDir, 'interaction_2d.png'));
close(fig3);

%% Save
timestamp = datestr(now, 'yyyymmdd_HHMMSS');
save(fullfile(outputDir, sprintf('sensitivity_analysis_%s.mat', timestamp)), ...
     'morrisRes', 'oatResults', 'paramNames', 'optimalParams', 'sortIdx');
fprintf('\nResults saved to: %s\n\n', outputDir);

%% -----------------------------------------------------------------------
%  Helper functions
%% -----------------------------------------------------------------------

function scores = evaluateInBatches(paramSets, gt_masks, gt_am, batchSz)
    n = length(paramSets);
    scores = zeros(n, 1);
    nB = ceil(n / batchSz);
    for b = 1:nB
        i1 = (b-1)*batchSz + 1;
        i2 = min(b*batchSz, n);
        sz = i2 - i1 + 1;
        tmp = zeros(sz, 1);
        parfor i = 1:sz
            tmp(i) = fitness_function(paramSets{i1+i-1}, gt_masks, gt_am);
        end
        scores(i1:i2) = tmp;
        fprintf('  batch %d/%d done\n', b, nB);
    end
end

function [scores, maes] = evaluateInBatchesFull(paramSets, gt_masks, gt_am, batchSz)
    n = length(paramSets);
    scores = zeros(n, 1); maes = zeros(n, 1);
    nB = ceil(n / batchSz);
    for b = 1:nB
        i1 = (b-1)*batchSz + 1;
        i2 = min(b*batchSz, n);
        sz = i2 - i1 + 1;
        tmpS = zeros(sz,1); tmpM = zeros(sz,1);
        parfor i = 1:sz
            [s, m] = fitness_function(paramSets{i1+i-1}, gt_masks, gt_am);
            tmpS(i) = s;
            tmpM(i) = m.MAE;
        end
        scores(i1:i2) = tmpS;
        maes(i1:i2)   = tmpM;
        fprintf('  batch %d/%d done\n', b, nB);
    end
end

function B = morrisTrajectory(p, delta)
    B_star = tril(ones(p+1, p), -1);
    P      = eye(p); P = P(randperm(p), :);
    D_star = diag(2*randi([0,1], 1, p) - 1);
    x_star = rand(1, p);
    J      = ones(p+1, p);
    B      = (J(1,:) .* x_star + delta * ((2*B_star - J)*D_star + J)) * P;
end

function d = trajectoryDistance(t1, t2)
    k = size(t1, 2);
    d = 0;
    for i = 1:(k+1)
        for j = 1:(k+1)
            d = d + sum((t1(i,:) - t2(j,:)).^2);
        end
    end
end

function sel = selectMaxSpread(distMat, r)
    M = size(distMat, 1);
    [~, mx] = max(distMat(:));
    [a, b]  = ind2sub([M,M], mx);
    sel     = [a, b];
    rest    = setdiff(1:M, sel);
    for iter = 3:r
        scores = arrayfun(@(c) sum(distMat(c, sel).^2), rest);
        [~, best] = max(scores);
        sel  = [sel, rest(best)];
        rest = setdiff(rest, rest(best));
    end
    sel = sort(sel);
end

function fname = sampleIDtoFilename(id)
    if id(1) == 'K', prefix = 'KRATS_'; else, prefix = 'XRATS_'; end
    fname = sprintf('%s%03d.nii.gz', prefix, str2double(id(2:end)));
end

File: python/sensitivity_analysis.py
"""sensitivity_analysis.py
Three-phase sensitivity analysis of the cyst-counting parameters.

  Phase 1 – Morris screening (SALib implementation)
  Phase 2 – One-at-a-time (OAT) analysis, all 6 parameters ± 30 %
  Phase 3 – 2-D parameter interaction for the top 2 parameters

This is a Python translation of sensitivity_analysis.m.
Phase 1 uses SALib instead of a manual trajectory implementation;
results are methodologically equivalent.
"""

import sys
import time
from pathlib import Path
from concurrent.futures import ProcessPoolExecutor
import os

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

sys.path.insert(0, str(Path(__file__).parent))
from fitness_function import evaluate_params

# ---- Configuration --------------------------------------------------------

OPTIMAL_PARAMS = np.array([
    4.27269938837114,
    18.2632731580751,
    42144.3492985436,
    14.9647743085524,
    0.111299008136904,
    0.918315428393029,
])
PARAM_NAMES = ['minPeakSepSmall', 'minPeakSepLarge', 'largeCystVolThresh',
               'maxPeakDistMerge', 'thresholdFraction', 'gaussianSigma']
LB = OPTIMAL_PARAMS * 0.7
UB = OPTIMAL_PARAMS * 1.3

BATCH_SIZE = 8


def _id_to_filename(sample_id: str) -> str:
    prefix = 'KRATS_' if sample_id[0] == 'K' else 'XRATS_'
    return f'{prefix}{int(sample_id[1:]):03d}.nii.gz'


def _eval_batch(param_sets, gt_masks, gt_counts, n_workers=None):
    """Evaluate a list of parameter sets; return (MAPE_array, MAE_array)."""
    if n_workers is None:
        n_workers = os.cpu_count() or 1
    scores = []
    maes   = []
    n  = len(param_sets)
    nB = (n + BATCH_SIZE - 1) // BATCH_SIZE
    for b in range(nB):
        batch = param_sets[b*BATCH_SIZE:(b+1)*BATCH_SIZE]
        for p in batch:
            s, m = evaluate_params(p, gt_masks, gt_counts, n_workers=n_workers)
            scores.append(s)
            maes.append(m['MAE'])
        print(f'  batch {b+1}/{nB} done')
    return np.array(scores), np.array(maes)


def run_sensitivity_analysis():
    repo_root  = Path(__file__).resolve().parent.parent
    data_dir   = repo_root / 'data' / 'Optimization set'
    mask_dir   = data_dir / 'Cyst masks'
    xls_file   = data_dir / 'O1_count.xlsx'
    output_dir = repo_root / 'results' / 'sensitivity_analysis'
    output_dir.mkdir(parents=True, exist_ok=True)

    if not xls_file.is_file():
        sys.exit(f'Ground-truth file not found:\n  {xls_file}')

    df = pd.read_excel(xls_file)
    gt_masks  = []
    gt_counts = []
    for _, row in df.iterrows():
        sid  = str(row.iloc[0])
        path = mask_dir / _id_to_filename(sid)
        if not path.is_file():
            sys.exit(f'Mask not found for {sid}:\n  {path}')
        gt_masks.append(str(path))
        gt_counts.append(int(row.iloc[1]))
    gt_counts = np.array(gt_counts)

    # ---- Phase 1: Morris screening ----------------------------------------
    print('\n=== Phase 1: Morris screening ===')
    t0 = time.time()

    try:
        from SALib.sample import morris as morris_sample
        from SALib.analyze import morris as morris_analyze
    except ImportError:
        sys.exit('SALib is required: pip install SALib')

    problem = {
        'num_vars': 6,
        'names':    PARAM_NAMES,
        'bounds':   list(zip(LB.tolist(), UB.tolist())),
    }

    # r=20 trajectories, matches MATLAB
    X_norm = morris_sample.sample(problem, N=20, num_levels=6,
                                  optimal_trajectories=None, seed=42)
    print(f'Morris sample: {X_norm.shape[0]} evaluations')

    param_sets = [X_norm[i] for i in range(X_norm.shape[0])]
    scores_m, _ = _eval_batch(param_sets, gt_masks, gt_counts, n_workers=1)

    Si = morris_analyze.analyze(problem, X_norm, scores_m, num_resamples=100, seed=42)

    mu_star = Si['mu_star']
    mu      = Si['mu']
    sigma   = Si['sigma']

    print(f'\nMorris completed in {(time.time()-t0)/60:.1f} min')
    print(f'\n{"Parameter":<25}  {"mu*":>8}  {"mu":>8}  {"sigma":>8}')
    print('-' * 55)
    sort_idx = np.argsort(mu_star)[::-1]
    thr = np.median(mu_star)
    for i in sort_idx:
        star = ' *' if mu_star[i] > thr else ''
        print(f'{PARAM_NAMES[i]:<25}  {mu_star[i]:>8.4f}  {mu[i]:>8.4f}  {sigma[i]:>8.4f}{star}')

    # Morris plot
    fig, ax = plt.subplots(figsize=(8, 6))
    ax.scatter(mu_star, sigma, s=100, alpha=0.8)
    for i, name in enumerate(PARAM_NAMES):
        ax.annotate(name, (mu_star[i]*1.03, sigma[i]*1.03), fontsize=9)
    ax.axvline(thr, color='r', ls='--', lw=1.5)
    ax.axhline(np.median(sigma), color='b', ls='--', lw=1.5)
    ax.set_xlabel('μ* (mean absolute elementary effect)', fontweight='bold')
    ax.set_ylabel('σ (standard deviation)', fontweight='bold')
    ax.set_title('Morris screening', fontweight='bold')
    ax.grid(True)
    fig.tight_layout()
    fig.savefig(output_dir / 'morris_screening.png', dpi=150)
    plt.close(fig)

    # ---- Phase 2: OAT analysis --------------------------------------------
    print('\n=== Phase 2: One-at-a-time analysis ===')
    t1 = time.time()

    variations = np.linspace(-0.3, 0.3, 21)
    oat_param_sets = []
    param_map      = []
    for rank, pi in enumerate(sort_idx):
        for v_idx, v in enumerate(variations):
            tp = OPTIMAL_PARAMS.copy()
            tp[pi] *= (1 + v)
            oat_param_sets.append(tp)
            param_map.append((rank, v_idx))

    print(f'Evaluating {len(oat_param_sets)} parameter combinations...')
    oat_scores, oat_maes = _eval_batch(oat_param_sets, gt_masks, gt_counts, n_workers=1)
    print(f'OAT completed in {(time.time()-t1)/60:.1f} min')

    oat_results = []
    for rank in range(len(sort_idx)):
        idxs = [i for i, (r, _) in enumerate(param_map) if r == rank]
        oat_results.append({
            'param':      PARAM_NAMES[sort_idx[rank]],
            'variations': variations,
            'scores':     oat_scores[idxs],
            'MAEs':       oat_maes[idxs],
        })

    fig, axes = plt.subplots(2, 3, figsize=(16, 9))
    for rank, (d, ax) in enumerate(zip(oat_results, axes.flat)):
        ax2 = ax.twinx()
        ax.plot(d['variations']*100, d['scores'], 'o-', lw=2, ms=6, color='#3366CC')
        ax2.plot(d['variations']*100, d['MAEs'],  's--', lw=2, ms=6, color='#CC3333')
        ax.set_xlabel('Variation (%)')
        ax.set_ylabel('MAPE (%)', color='#3366CC')
        ax2.set_ylabel('MAE (cysts)', color='#CC3333')
        ax.axvline(0, color='r', ls='--', lw=2)
        ax.set_title(f'{d["param"]} (rank {rank+1})', fontweight='bold')
        ax.grid(True)
    fig.suptitle('One-at-a-time sensitivity analysis', fontsize=14, fontweight='bold')
    fig.tight_layout()
    fig.savefig(output_dir / 'oat_analysis.png', dpi=150)
    plt.close(fig)

    # ---- Phase 3: 2-D interaction -----------------------------------------
    print('\n=== Phase 3: 2-D parameter interaction ===')
    t2 = time.time()

    p1_idx = sort_idx[0]
    p2_idx = sort_idx[1]
    grid_size = 9
    grid_vals = np.linspace(-0.25, 0.25, grid_size)

    grid_sets = []
    grid_map  = []
    for i, vi in enumerate(grid_vals):
        for j, vj in enumerate(grid_vals):
            tp = OPTIMAL_PARAMS.copy()
            tp[p1_idx] *= (1 + vi)
            tp[p2_idx] *= (1 + vj)
            grid_sets.append(tp)
            grid_map.append((i, j))

    print(f'Computing {grid_size}×{grid_size} interaction grid...')
    g_scores, g_maes = _eval_batch(grid_sets, gt_masks, gt_counts, n_workers=1)
    print(f'2-D interaction completed in {(time.time()-t2)/60:.1f} min')

    int_scores = np.zeros((grid_size, grid_size))
    int_maes   = np.zeros((grid_size, grid_size))
    for k, (i, j) in enumerate(grid_map):
        int_scores[i, j] = g_scores[k]
        int_maes[i, j]   = g_maes[k]

    fig, axes = plt.subplots(1, 2, figsize=(13, 5))
    for ax, data, title in [(axes[0], int_scores, 'MAPE landscape (%)'),
                             (axes[1], int_maes,   'MAE landscape (cysts)')]:
        im = ax.imshow(data, extent=[grid_vals[0]*100, grid_vals[-1]*100,
                                     grid_vals[-1]*100, grid_vals[0]*100],
                       aspect='auto', origin='upper', cmap='jet')
        plt.colorbar(im, ax=ax)
        ax.set_xlabel(f'{PARAM_NAMES[p2_idx]} variation (%)', fontweight='bold')
        ax.set_ylabel(f'{PARAM_NAMES[p1_idx]} variation (%)', fontweight='bold')
        ax.set_title(title, fontweight='bold')
        ax.plot(0, 0, 'wx', ms=15, mew=3)
    fig.suptitle(f'Parameter interaction: {PARAM_NAMES[p1_idx]} vs {PARAM_NAMES[p2_idx]}',
                 fontsize=13, fontweight='bold')
    fig.tight_layout()
    fig.savefig(output_dir / 'interaction_2d.png', dpi=150)
    plt.close(fig)

    # ---- Save results ------------------------------------------------------
    from datetime import datetime
    ts = datetime.now().strftime('%Y%m%d_%H%M%S')
    np.savez(output_dir / f'sensitivity_analysis_{ts}.npz',
             mu_star=mu_star, mu=mu, sigma=sigma,
             sort_idx=sort_idx, param_names=PARAM_NAMES,
             optimal_params=OPTIMAL_PARAMS)
    print(f'\nResults saved to: {output_dir}\n')


if __name__ == '__main__':
    run_sensitivity_analysis()
